rules/namespaces.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/m3db/m3metrics/generated/proto/schema"
+	"github.com/m3db/m3metrics/rules"
+	"github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/log"
+)
+
+var (
+	emptyNamespaces rules.Namespaces
+	errNilValue     = errors.New("nil value received")
+)
+
+// namespaces contains the list of namespace users have defined
+// rules for. The namespace list is updated as needed.
+type namespaces struct {
+	sync.RWMutex
+	runtime.Value
+
+	key          string
+	store        kv.Store
+	cache        Cache
+	opts         Options
+	log          xlog.Logger
+	ruleSetKeyFn RuleSetKeyFn
+
+	proto *schema.Namespaces
+	rules map[string]*ruleSets
+}
+
+func newNamespaces(key string, cache Cache, opts Options) *namespaces {
+	n := &namespaces{
+		key:          key,
+		store:        opts.KVStore(),
+		cache:        cache,
+		opts:         opts,
+		log:          opts.InstrumentOptions().Logger(),
+		ruleSetKeyFn: opts.RuleSetKeyFn(),
+		proto:        &schema.Namespaces{},
+		rules:        make(map[string]*ruleSets),
+	}
+	valueOpts := runtime.NewOptions().
+		SetClockOptions(opts.ClockOptions()).
+		SetInstrumentOptions(opts.InstrumentOptions()).
+		SetInitWatchTimeout(opts.InitWatchTimeout()).
+		SetKVStore(n.store).
+		SetUnmarshalFn(n.toNamespaces).
+		SetVersionFn(namespacesVersion).
+		SetCutoverFn(namespacesCutover).
+		SetProcessFn(n.process)
+	n.Value = runtime.NewValue(key, valueOpts)
+	return n
+}
+
+func (n *namespaces) Close() error {
+	multiErr := xerrors.NewMultiError()
+	if err := n.Value.Close(); err != nil {
+		multiErr = multiErr.Add(err)
+	}
+
+	n.RLock()
+	defer n.RUnlock()
+	for _, rs := range n.rules {
+		if err := rs.Close(); err != nil {
+			multiErr = multiErr.Add(err)
+		}
+	}
+	return multiErr.FinalError()
+}
+
+func (n *namespaces) toNamespaces(value kv.Value) (interface{}, error) {
+	if value == nil {
+		return emptyNamespaces, errNilValue
+	}
+	n.proto.Reset()
+	if err := value.Unmarshal(n.proto); err != nil {
+		return emptyNamespaces, err
+	}
+	return rules.NewNamespaces(n.proto), nil
+}
+
+func (n *namespaces) process(value interface{}) {
+	var (
+		nss        = value.(rules.Namespaces)
+		namespaces = nss.Namespaces()
+		incoming   = make(map[string]struct{}, len(namespaces))
+		added      = make(map[string]*ruleSets)
+	)
+	for _, ns := range namespaces {
+		incoming[ns] = struct{}{}
+	}
+
+	n.RLock()
+	// NB(xichen): if a namespace is deleted, we don't remove it from
+	// the map right away because there may be rules in it that don't
+	// become invalid until certain time in the future. Therefore we
+	// simply mark the rulesets for that namespace deleted and only
+	// remove it from the rules map when the rules have expired.
+	for ns, existing := range n.rules {
+		if _, ok := incoming[ns]; !ok {
+			if err := existing.MarkTombstoned(); err != nil {
+				n.log.WithFields(
+					xlog.NewLogField("ruleSetKey", existing.Key()),
+					xlog.NewLogErrField(err),
+				).Error("failed to mark ruleset as tombstoned")
+			}
```

Comment:
Perhaps a continue at the end of the `if` clause, to avoid the `else` clause and the indentation below.